### PR TITLE
Check memory for OOM

### LINKED ISSUE #2461: OutOfMemory exception on opening big text files in editor
## Exception
* User Action: UI Error
* Request: Application crash
* Version: 3.5.3
* OS: Linux samsung/crownltexx/crownlte:10/QP1A.190711.020/N960FXXU6FTK1:user/release-keys 10 - 29
* Device: crownlte
* Model: SM-N960F
* Product: crownltexx
<details><summary><b>Crash log </b></summary><p>

java.lang.RuntimeException: An error occurred while executing doInBackground()
 at android.os.AsyncTask$4.done(AsyncTask.java:399)
 at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
 at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
 at java.util.concurrent.FutureTask.run(FutureTask.java:271)
 at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
 at java.lang.Thread.run(Thread.java:919)
Caused by: java.lang.OutOfMemoryError: Failed to allocate a 174063624 byte allocation with 25149440 free bytes and 165MB until OOM, target footprint 119906520, growth limit 268435456
 at java.util.Arrays.copyOf(Arrays.java:3257)
 at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
 at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
 at java.lang.StringBuilder.append(StringBuilder.java:137)
 at com.amaze.filemanager.asynchronous.asynctasks.ReadFileTask.doInBackground(ReadFileTask.java:118)
 at com.amaze.filemanager.asynchronous.asynctasks.ReadFileTask.doInBackground(ReadFileTask.java:47)
 at android.os.AsyncTask$3.call(AsyncTask.java:378)
 at java.util.concurrent.FutureTask.run(FutureTask.java:266)
 ... 4 more

</details>
<hr>

### PR DESCRIPTION
**Pull request by VishnuSanal**
#### Issue tracker   

Fixes #2461 

  
#### Manual tests
- [x] Done  
  

#### Build tasks success  

Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java [hunk 1]
```diff
@@ -633,4 +646,22 @@ private void cleanSpans() {
       mInput.getText().removeSpan(colorSpan);
     }
   }
+
+  private boolean isMemoryAvailable(long size) {
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/activities/TextEditorActivity.java:650]: You also need on low memory.

File: app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java [hunk 2]
```diff
@@ -126,6 +127,9 @@ protected ReturnedValues doInBackground(Void... params) {
     } catch (IOException e) {
       e.printStackTrace();
       return new ReturnedValues(EXCEPTION_IO);
+    } catch (OutOfMemoryError e) {
+      e.printStackTrace();
+      return new ReturnedValues(EXCEPTION_OOM);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java:132]: How will it create a new object if it has no more memory?

Comment (VishnuSanal) [reply app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java:132]: > How will it create a new object if it has no more memory?

How else? Any suggestions?

Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java:132]: * You can check if you have enough memory _before_ trying to reserve it.
* You can use a window based system to access the file intelligently (I know that search is an issue, but you can simply disable search for now if you find that using another window to search is too complicated).
* You can use a `malloc` like system that allows for a "null" response if you can't allocate the memory, and that basically checks if you have the memory before trying to allocate.

Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/ReadFileTask.java:132]: * You can load memory in `chunks` and stop if onLowMemory or onTrimMemory are called.

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**